File: templates/oc/Example/PROJECT/Modules/Main/TabBarController.m
//
//  TabBarController.m
//  PROJECT
//
//  Created by USER_NAME on TODAYS_DATE.
//

#import "TabBarController.h"

#import <UIColor+YYAdd.h>

#import "NavigationController.h"

#import "CPDHomeViewController.h"
#import "CPDProfileViewController.h"

@implementation UIViewController (TabBarControllerAdded)

- (void)tabbarItemDidSelect {}
- (void)tabbarItemDidDeselect {}
- (void)tabbarItemDidSelectAgain {}

@end

@interface TabBarController ()
<UITabBarControllerDelegate>

@property (nonatomic, unsafe_unretained) UIViewController *lastSelectedVC;
@property (nonatomic, assign) NSTimeInterval lastSelectedTimestamp;
@end

@implementation TabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.delegate = self;
    [self addChildVCWithClass:[CPDHomeViewController class] iconName:@"tabbar_home" title:@"首页"];
    [self addChildVCWithClass:[CPDProfileViewController class] iconName:@"tabbar_profile" title:@"我"];
}


- (__kindof UIViewController *)addChildVCWithClass:(Class)cls iconName:(NSString *)iconName title:(NSString *)title {
    UIViewController *vc = [cls new];
    vc.tabBarItem.image = [UIImage imageNamed:iconName];
    vc.tabBarItem.selectedImage = [UIImage imageNamed:[iconName stringByAppendingString:@"_selected"]];
    UIFont *font = [UIFont systemFontOfSize:11];
    
    [vc.tabBarItem setTitleTextAttributes:@{NSFontAttributeName: font, NSForegroundColorAttributeName: [UIColor colorWithHexString:@"0x6D6D6D"]} forState:UIControlStateNormal];
    [vc.tabBarItem setTitleTextAttributes:@{NSFontAttributeName: font, NSForegroundColorAttributeName: [UIColor colorWithHexString:@"0x696969"]} forState:UIControlStateSelected];
    
    vc.title = title;
    [self addChildViewController:[[NavigationController alloc] initWithRootViewController:vc]];
    return vc;
}

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController {
    
    UIViewController *(^findOp)(void) = ^{
        if ([viewController isKindOfClass:[UINavigationController class]]) {
            UINavigationController *nav = (UINavigationController *)viewController;
            return nav.topViewController;
        }
        return viewController;
    };
    UIViewController *topVC = findOp();
    NSTimeInterval now = [NSDate date].timeIntervalSinceReferenceDate;
    if (_lastSelectedVC == topVC) {
        if (now - _lastSelectedTimestamp < 0.5) {
            [topVC tabbarItemDidSelectAgain];
        }
    } else {
        [topVC tabbarItemDidSelect];
        [_lastSelectedVC tabbarItemDidDeselect];
    }
    _lastSelectedVC = topVC;
    _lastSelectedTimestamp = now;
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: templates/swift/Example/PROJECT/Modules/Main/TabBarController.swift
//
//  TabBarController.swift
//  PROJECT
//
//  Created by USER_NAME on TODAYS_DATE.
//

import UIKit
 

class TabBarController: UITabBarController {
    
    private var lastSelectedVC: UIViewController?
    private var lastSelectedTimestamp: TimeInterval = 0
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.delegate = self
        addChild(HomeViewController.self, "tabbar_home", "首页")
        addChild(ProfileViewController.self, "tabbar_profile", "我")
    }
    
    @discardableResult
    private func addChild<T: UIViewController>(_ viewController: T.Type,
                                               _ image: String,
                                               _ title: String) -> T {
        let vc = T()
        vc.tabBarItem.image = UIImage(named: image)
        vc.tabBarItem.selectedImage = UIImage(named: image + "_selected")
        let font = UIFont.systemFont(ofSize: 11)
        vc.tabBarItem.setTitleTextAttributes([
            .font: font,
            .foregroundColor: UIColor(gray: 105)], for: .normal)
        vc.tabBarItem.setTitleTextAttributes([
            .font: font,
            .foregroundColor: UIColor(gray: 105)], for: .selected)
        vc.title = title
        addChild(NavigationController(rootViewController: vc))
        return vc
    }
    
}
extension TabBarController: UITabBarControllerDelegate {
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        let topVC: UIViewController? = {
            if let nav = viewController as? UINavigationController {
                return nav.topViewController
            }
            return viewController
        }()
        
        let now = Date().timeIntervalSinceReferenceDate
        if lastSelectedVC === topVC {
            if now - lastSelectedTimestamp < 0.5 { topVC?.tabbarButtonItemDidSelectAgain()
            }
        } else {
            topVC?.tabbarButtonItemDidSelect()
            lastSelectedVC?.tabbarButtonItemDidDeselect()
        }
        lastSelectedVC = topVC
        lastSelectedTimestamp = now
    }
}


extension UIViewController {
    @objc func tabbarButtonItemDidSelect() {}
    @objc func tabbarButtonItemDidDeselect() {}
    @objc func tabbarButtonItemDidSelectAgain() {}
}
